### PR TITLE
Test enhancements

### PR DESCRIPTION
**Pull request by jmonteer**
WIP

Working on improving the tools that a strategist have to test a new strategy. 
The objective is (i) reduce time that we spend setting up tests, (ii) test strategies in a more comprehensive way (iii) ...

This will include:

- [x] more detailed check list of tests to be done
- [x] more detailed templates for those tests than can be thought of in advance
- [x] tests run isolated in a fresh fork every time
- [x] utils like unit converters, standard debugging prints or standard checks to be done after certain actions
- [ ] improvement of the time to run tests to avoid lengthy waits that make strategists waste time while writing tests
- [x] parametrized tests for testing different tokens (or other scenarios) using the same tests
- [ ] integration with live vaults to run tests in production
- [ ] integration with Tenderly to improve debugging

Please add any other suggestions or requests

Linear ticket: 
https://linear.app/yearn/issue/BRN-75/test-suite-enhancements

### CODE REVIEW

**Comment by jmonteer** [comment]
I will most probably open a new branch for a deeper than initially thought brownie <> tenderly integration

**Comment by jmonteer** [comment]
So far: 

- added healthcheck support using -1 dependency
- added `strat_ms` fixture to act from strategists ms
- parametrized token fixture and added a list of supported tokens (only including those that are most important atm)
- modified fixture "amount" to be a deposit of around $1m, independently of the token
- created a fixture with whales for each token used
- added a clean fixture to snapshot and revert to a clean state before each specific test
- added a couple of fixtures regarding registry, to support the automation of getting latest production vault for each token
- added a fixture to get a cloned instance of the strategy being tested
- added a fixture to always withdraw after each test, to be sure that withdrawing is possible in any situation, unless explicitely avoided
- added a test_airdrop as requested in #29 
- added a test_clone to test that clones are correctly created and are harvestable
- modified test_harvest to add a case with both profits and losses, to ensure accounting is working
- created utils suite with 
       a) actions: repetitive actions like deposit in a vault from a user
       b) checks: standard checks with asserts
       c) utils: unit converters or sleeps, to avoid repetitive code and improve readability
- added a test file for manual functions (for those strategies that implement them)
- made a bit more detailed test for test_migration
- added a test for restricted functions so strategist can be sure they thought through the correct set up
- added a test for test_revoke_with_profit to be sure that even when profit is added to the strategy, it is revokable (we've had problems in the past with this)
- added a test for a strategy unwind


Tenderly network management and debugging might be included in another PR, depending on timing

It would be nice to have your comments @salazarguille @storming0x @Grandthrax @poolpitako @fameal

**Comment by salazarguille** [comment]
Great job @jmonteer! qq, are you planning to add unit tests to cover different scenarios with the health check?

**Comment by jmonteer** [comment]
yeah would love to add those. still have to do some catch up with healthcheck tho

**Comment by jmonteer** [comment]
@salazarguille added test_healtcheck 

please add any other ideas you feel would be useful!

File: tests/conftest.py [hunk 1]
```diff
@@ -38,19 +49,71 @@ def keeper(accounts):
...
-def token():
-    token_address = "0x6b175474e89094c44da98b954eedeac495271d0f"  # this should be the address of the ERC-20 used by the strategy/vault (DAI)
-    yield Contract(token_address)
-
-
-@pytest.fixture
-def amount(accounts, token, user):
-    amount = 10_000 * 10 ** token.decimals()
+token_addresses = {
+    "WBTC": "0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599",  # WBTC
```
Comment (stcrab) [inline tests/conftest.py:53]: Is that relevant to have some tokens addresses already here?

File: tests/conftest.py [hunk 2]
```diff
@@ -38,19 +49,71 @@ def keeper(accounts):
...
+}
+
+
+@pytest.fixture(scope="session", autouse=True)
+def token_whale(token):
+    yield whale_addresses[token.symbol()]
+
+
+token_prices = {
+    "WBTC": 35_000,
```
Comment (stcrab) [inline tests/conftest.py:97]: Could be fetched in a batch using the coingeko API so that remains up to date.

https://www.coingecko.com/en/api

File: tests/conftest.py [hunk 3]
```diff
@@ -85,6 +158,31 @@ def strategy(strategist, keeper, vault, Strategy, gov):
...
+        strategist, {"from": strategist}
+    ).return_value
+    cloned_strategy = Strategy.at(cloned_strategy)
+    vault.revokeStrategy(strategy)
+    vault.addStrategy(cloned_strategy, 10_000, 0, 2 ** 256 - 1, 1_000, {"from": gov})
+    yield
+
+
+@pytest.fixture(autouse=True)
+def withdraw_no_losses(vault, token, amount, user):
```
Comment (stcrab) [inline tests/conftest.py:175]: Doesn't seems to be used.

File: tests/test_harvests.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+# tests harvesting a strategy that returns profits correctly
+def test_profitable_harvest(
+    chain, accounts, token, vault, strategy, user, strategist, amount, RELATIVE_APPROX
+):
+    # Deposit to the vault
+    actions.user_deposit(user, vault, token, amount)
+
+    # Harvest 1: Send funds through the strategy
+    chain.sleep(1)
```
Comment (stcrab) [inline tests/test_harvests.py:12]: Maybe a helper can be added to do:

```
    chain.sleep(1)
    strategy.harvest({"from": strategist})
```

File: tests/test_healthcheck.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    # TODO: generate a unacceptable loss
+    loss_amount = amount * 0.05
+    actions.generate_loss(loss_amount)
+
+    # Harvest should revert because the loss in unacceptable
+    with brownie.reverts("!healthcheck"):
+        strategy.harvest({"from": strategist})
+
+    # we disable the healthcheck
+    strategy.setDoHealthCheck(False, {"from": gov})
```
Comment (stcrab) [inline tests/test_healthcheck.py:30]: will break on 0.4.3, checks are on the vault

File: tests/conftest.py [hunk 6]
```diff
@@ -38,19 +49,71 @@ def keeper(accounts):
...
+    ],
+    scope="session",
+    autouse=True,
+)
+def token(request):
+    yield Contract(token_addresses[request.param])
+
+
+whale_addresses = {
+    "WBTC": "0x28c6c06298d514db089934071355e5743bf21d60",
```
Comment (stcrab) [inline tests/conftest.py:81]: Found this API you can use to fetch token holders.

https://github.com/EverexIO/Ethplorer/wiki/Ethplorer-API#get-top-token-holders

File: tests/conftest.py (hunk 2 above)
Comment (jmonteer) [reply tests/conftest.py:97]: wouldn't use it. only the order of magnitude is important, otherwise tests might break with different prices

File: tests/test_harvests.py (hunk 4 above)
Comment (jmonteer) [reply tests/test_harvests.py:12]: agree

File: tests/conftest.py (hunk 6 above)
Comment (jmonteer) [reply tests/conftest.py:81]: this one is useful. will look into it

**Review by storming0x — COMMENTED** [review]
Great effort, do we need to update this to work with 0.4.3?
